### PR TITLE
[AOTI] Add C shim for _mkl_linear

### PR DESCRIPTION
**Pull request by chunyuan-w**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137880

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/csrc/inductor/aoti_torch/c/shim_mkldnn.h [hunk 1]
```diff
@@ -129,6 +129,18 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_cpu__linear_pointwise_binary(
     const char* attr,
     AtenTensorHandle* ret0);
 
+#if AT_MKL_ENABLED()
```
Comment (desertfire) [inline torch/csrc/inductor/aoti_torch/c/shim_mkldnn.h:132]: Same here for the nested macro question

File: torch/csrc/inductor/aoti_torch/shim_mkldnn.cpp [hunk 2]
```diff
@@ -269,4 +269,26 @@ AOTITorchError aoti_torch_cpu__linear_pointwise_binary(
   });
 }
 
+#if AT_MKL_ENABLED()
```
Comment (desertfire) [inline torch/csrc/inductor/aoti_torch/shim_mkldnn.cpp:272]: and here

File: aten/src/ATen/native/mkldnn/Linear.h [hunk 3]
```diff
@@ -22,6 +22,17 @@ C10_API Tensor mkldnn_linear_pointwise_binary(
     const std::optional<Tensor>& bias_opt,
     c10::string_view attr);
 
+#if AT_MKL_ENABLED()
```
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply aten/src/ATen/native/mkldnn/Linear.h:25]: I followed the nested macro here: https://github.com/pytorch/pytorch/blob/906fe058958974512d6fab6aba83aece28e82ca3/aten/src/ATen/native/mkldnn/Linear.cpp#L339 is inside of https://github.com/pytorch/pytorch/blob/906fe058958974512d6fab6aba83aece28e82ca3/aten/src/ATen/native/mkldnn/Linear.cpp#L52.

In the `mkl_linear` function, we need MKLDNN utility functions for example https://github.com/pytorch/pytorch/blob/906fe058958974512d6fab6aba83aece28e82ca3/aten/src/ATen/native/mkldnn/Linear.cpp#L387 thus it is defined inside of `AT_MKLDNN_ENABLED`.

File: torch/csrc/inductor/aoti_torch/c/shim_mkldnn.h (hunk 1 above)
Comment (chunyuan-w) [reply torch/csrc/inductor/aoti_torch/c/shim_mkldnn.h:132]: ditto

File: torch/csrc/inductor/aoti_torch/shim_mkldnn.cpp (hunk 2 above)
Comment (chunyuan-w) [reply torch/csrc/inductor/aoti_torch/shim_mkldnn.cpp:272]: ditto

**Review by desertfire — APPROVED** [review]

**Review by desertfire — COMMENTED** [review]
You need to rebase after https://github.com/pytorch/pytorch/pull/138047

File: torch/_inductor/mkldnn_ir.py [hunk 4]
```diff
@@ -1095,16 +1095,8 @@ def __init__(
...
-            self.python_kernel_name,
-            self.cpp_kernel_name,
-            self.codegen_args(),
-            self.cpp_op_schema,
-            self.cpp_kernel_key,
-            op_overload=self.op_overload,
-            raw_args=[*self.inputs, *self.constant_args],
-        )
+        wrapper.include_extra_header("torch/csrc/inductor/aoti_torch/c/shim_mkldnn.h")
+        super().codegen(wrapper)
```
Comment (chunyuan-w) [inline torch/_inductor/mkldnn_ir.py:1099]: Remove non-ABI-compatible mode following https://github.com/pytorch/pytorch/pull/138047

**Pull request closed without merging**